송수경/4월 4주차/BOJ_G2_1938_통나무옮기기.java

```diff
@@ -0,0 +1,127 @@
+package boj.Gold;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.LinkedList;
+import java.util.Queue;
+
+public class BOJ_G2_1938_통나무옮기기 {
+	static char[][] map;
+	static int N, sr, sc, sf, er, ec, ef;	//크기, 시작점, 최종점, 최종점 가로(1) 혹은 세로(2) 상태
+	static Queue<Tree> q = new LinkedList<Tree>();
+	
+	static class Tree {
+		int r, c, d, f;	//위치, 움직임 횟수, 보는 방향(가로 1, 세로 2)
+		
+		public Tree(int r, int c, int d, int f) {
+			this.r = r;
+			this.c = c;
+			this.d = d;
+			this.f = f;
+		}
+	}
+	
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+		N = Integer.parseInt(br.readLine().trim());	//평지의 한 변의 길이
+		map = new char[N][N];
+		
+//		1. 입력받기(최초 위치, 최종 위치)
+		for(int i=0, start=0, end=0; i<N; i++) {
+			map[i] = br.readLine().toCharArray();
+			for(int j=0; j<N; j++) {
+				if(map[i][j]=='B') {	//B를 두번째 만났을 때 가로/세로 판별
+					if(start==0) { sr=i; sc=j; start++;}
+					else if(start++==1) {
+						sf = i-sr+1; sr=i; sc=j;
+						q.offer(new Tree(sr, sc, 0, sf));
+					}
```

Comment:
이부분은 시작점 배열과 최종점 배열을 만들어서 일단 전부 입력받고 나중에 좌표간의 차로 수평, 수직을 판별하면 메모리 낭비가 덜할것 같아요!

> ``
```suggestion
	        static int[][] target, pos;
		for(int i=0; i<N; i++) {
			String line = in.readLine();
			for(int j=0; j<N; j++) {
				map[i][j] = line.charAt(j);
				if(map[i][j] == 'B') {
					pos[bidx][0] = i;
					pos[bidx++][1] = j;
				}else if(map[i][j] == 'E') {
					target[eidx][0] = i;
					target[eidx++][1] = j;
				}
			}
		}
		tdir = target[0][0] - target[1][0] == 0 ? 0 : 1; // r의 차가 0이면 수평, 1이면 수직
```

Reply:
@choieunsong 아하 비교 횟수때문인가요??